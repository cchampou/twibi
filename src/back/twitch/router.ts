import {
  callback,
  createSubscription,
  listSubscriptions,
  revokeSubscription,
} from './controllers/events';
import {
  clearCommands,
  createCommand,
  listCommands,
} from './controllers/commands';
import { login } from './controllers/auth';
import {
  get,
  subscribe,
  subscribeHost,
  unsubscribe,
  unsubscribeHost,
} from './controllers/notification';
import authMiddleware from '../core/auth.middleware';

const express = require('express');

const router = express.Router();

router.post('/login', login);

// Notification subscribe
router.use('/subscribe', authMiddleware);

router.get('/subscribe/:type', get);

router.post('/subscribe/host', subscribeHost);
router.post('/subscribe/:type', subscribe);

router.delete('/subscribe/host', unsubscribeHost);
router.delete('/subscribe/:type', unsubscribe);

// Events
router.get('/event', listSubscriptions);
router.post('/event', createSubscription);
router.delete('/event', revokeSubscription);
router.post('/event/callback', callback);

// Chatbot
router.post('/commands', createCommand);
router.get('/commands', listCommands);
router.delete('/commands', clearCommands);

export default router;
