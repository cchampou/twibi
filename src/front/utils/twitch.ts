/**
 * Generate twitch OAuth link
 * @param {string} clientId
 * @returns {string} url
 */
// eslint-disable-next-line import/prefer-default-export
export const generateTwitchOAuthLink = (clientId) =>
  `https://id.twitch.tv/oauth2/authorize?client_id=${clientId}&redirect_uri=${process.env.PUBLIC_URL}&response_type=token&scope=user:read:email chat:read chat:edit channel:read:subscriptions`;
