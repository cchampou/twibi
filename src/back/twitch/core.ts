import fetch from 'node-fetch';
import Auth from './auth';
import { logInfo, logWarn } from '../utils/logger';

// eslint-disable-next-line import/prefer-default-export
export const subscribeToWebhook: (topic: string) => void = (topic) => {
  const EventRequestParams = new URLSearchParams();
  EventRequestParams.append('hub.mode', 'subscribe');
  EventRequestParams.append('hub.topic', topic);
  EventRequestParams.append('hub.callback', 'http://176.190.176.219:8000/twitch/event/callback');
  EventRequestParams.append('hub.lease_seconds', '60');
  EventRequestParams.append('hub.secret', 'original');
  fetch('https://api.twitch.tv/helix/webhooks/hub', {
    method: 'POST',
    body: EventRequestParams,
    headers: Auth.generateHeaders(),
  })
    .then((res) => {
      logInfo(res.status.toString());
      return res.text();
    })
    .then((text) => {
      if (text.length) {
        logWarn(text);
      }
    });
};
