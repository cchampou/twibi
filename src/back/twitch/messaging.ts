import { Client } from 'tmi.js';
import { logError, logInfo } from '../utils/logger';
import { splitWords, trimStart } from '../utils/string';
import Database from '../core/database';

class TwitchMessaging {
  client;

  constructor() {
    this.client = new Client({
      identity: {
        username: process.env.TWITCH_BOT_USERNAME,
        password: process.env.TWITCH_BOT_PASSWORD,
      },
      channels: ['#carorockwell'],
    });
    this.client.connect().then(logInfo).catch(logError);
    this.client.on('message', this.onMessage);
    this.client.on('hosted', this.onHosted);
  }

  onHosted = (channel, username) => {
    // eslint-disable-next-line no-console
    console.log(channel, username);
    this.client.say('#carorockwell', `@${username} thank you for the host !`).catch(logInfo);
  };

  onMessage = (channel, tags, message, self) => {
    if (self) return;
    const words = splitWords(message);
    if (words[0][0] === '!') {
      this.commands(channel, tags, words);
    }
  };

  commands(channel, tags, words) {
    const found = this.findCommand(words);
    if (found) {
      const { response, data } = found;
      data.push(['{me}', tags.username]);
      this.client.say(channel, this.populate(response, data));
    }
  }

  populate(response, data) {
    if (data.length === 0) {
      return response;
    }
    return data.reduce((acc, val) => acc.replace(new RegExp(val[0], 'g'), trimStart(val[1], '@').toLowerCase()), response);
  }

  findCommand(words): { data:Array<[string, string]>, response: string } {
    const needle = trimStart(words[0], '!');
    const commands = Database.db.get('commands').value();
    const matchingPattern = commands.find(([command]) => command.startsWith(needle));
    if (!matchingPattern) return null;
    const wordedPattern = splitWords(matchingPattern[0]);
    return {
      data: wordedPattern
        .map((val: string, index: number): [string, string] => [val, words[index] || '']).slice(1),
      response: matchingPattern[1],
    };
  }
}

export default new TwitchMessaging();
