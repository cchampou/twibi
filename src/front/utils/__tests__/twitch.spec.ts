import { generateTwitchOAuthLink } from '../twitch';

describe('generateTwitchOAuthLink', () => {
  it('should match snapshot', () => {
    expect(generateTwitchOAuthLink('client-id')).toMatchSnapshot();
  });
});
