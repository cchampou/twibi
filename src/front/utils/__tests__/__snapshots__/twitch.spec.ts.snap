// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTwitchOAuthLink should match snapshot 1`] = `"https://id.twitch.tv/oauth2/authorize?client_id=client-id&redirect_uri=undefined&response_type=token&scope=user:read:email chat:read chat:edit channel:read:subscriptions"`;
